File: overlay_ws/src/agent_action/scripts/execute_motion_client.py
#!/usr/bin/env python3
import rclpy
from rclpy.action import ActionClient
from rclpy.node import Node

from std_msgs.msg import Float64MultiArray
from agent_action_interface.action import ExecuteMotion


class MotionClient(Node):

    def __init__(self):
        super().__init__("motion_action_client")

        # Action Client
        self._client = ActionClient(self, ExecuteMotion, "execute_motion")

        # Subscription
        self.subscription = self.create_subscription(
            Float64MultiArray,
            "published_coordinates",
            self.agent_callback,
            10,
        )

        self._goal_in_flight = False
        self.get_logger().info(
            "Motion action client initialised and listening to 'published_coordinates'..."
        )

    def agent_callback(self, msg: Float64MultiArray):
        # Avoid sending multiple goals at once
        if self._goal_in_flight:
            self.get_logger().warn("Goal already in flight. Ignoring new coordinates.")
            return

        data = list(msg.data)
        self.get_logger().info(f"Received topic goal: {data}")

        goal_msg = ExecuteMotion.Goal()
        goal_msg.data = data

        self._client.wait_for_server()
        self._goal_in_flight = True

        send_future = self._client.send_goal_async(
            goal_msg,
            feedback_callback=self.feedback_callback,
        )
        send_future.add_done_callback(self.goal_response_callback)

    def goal_response_callback(self, future):
        goal_handle = future.result()

        if not goal_handle.accepted:
            self.get_logger().error("Goal rejected")
            self._goal_in_flight = False
            return

        self.get_logger().info("Goal accepted")

        result_future = goal_handle.get_result_async()
        result_future.add_done_callback(self.result_callback)

    def feedback_callback(self, feedback_msg):
        fb = feedback_msg.feedback
        self.get_logger().info(
            f"Feedback: state={fb.state}, progress={fb.progress}"
        )

    def result_callback(self, future):
        try:
            result = future.result().result
            self.get_logger().info(
                f"Result: success={result.success}, "
                f"code={result.error_code}, "
                f"description={result.error_description}"
            )
        except Exception as e:
            self.get_logger().error(f"Failed to get result: {e}")
        finally:
            self._goal_in_flight = False


def main(args=None):
    rclpy.init(args=args)
    node = MotionClient()

    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()

File: overlay_ws/src/agent_action/src/execute_motion_client.cpp
#include <rclcpp/rclcpp.hpp>
#include <rclcpp_action/rclcpp_action.hpp>

#include "agent_action_interface/action/execute_motion.hpp"

class MotionClient : public rclcpp::Node
{
public:
    using ExecuteMotion = agent_action_interface::action::ExecuteMotion;
    using GoalHandleMotion = rclcpp_action::ClientGoalHandle<ExecuteMotion>;

    MotionClient()
        : Node("motion_action_client")
    {
        client_ = rclcpp_action::create_client<ExecuteMotion>(this, "execute_motion");
    }

    void send_goal(const std::vector<double>& data)
    {
        if (!client_->wait_for_action_server()) {
            RCLCPP_ERROR(this->get_logger(), "Action server not available");
            return;
        }

        auto goal_msg = ExecuteMotion::Goal();
        goal_msg.data = data;

        auto send_goal_options =
            rclcpp_action::Client<ExecuteMotion>::SendGoalOptions();

        send_goal_options.feedback_callback =
            std::bind(&MotionClient::feedback_callback, this, std::placeholders::_1, std::placeholders::_2);

        send_goal_options.result_callback =
            std::bind(&MotionClient::result_callback, this, std::placeholders::_1);

        client_->async_send_goal(goal_msg, send_goal_options);
    }

private:
    void feedback_callback(
        GoalHandleMotion::SharedPtr,
        const std::shared_ptr<const ExecuteMotion::Feedback> feedback)
    {
        RCLCPP_INFO(
            this->get_logger(),
            "Feedback: %s (%.2f)",
            feedback->state.c_str(),
            feedback->progress
        );
    }

    void result_callback(const GoalHandleMotion::WrappedResult &result)
    {
        RCLCPP_INFO(
            this->get_logger(),
            "Result: success=%d, code=%s, desc=%s",
            result.result->success,
            result.result->error_code.c_str(),
            result.result->error_description.c_str()
        );
    }

    rclcpp_action::Client<ExecuteMotion>::SharedPtr client_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<MotionClient>();

    node->send_goal({1, 0.4, 0.1, 0.3, 0.707, 0, -0.7070, 1});

    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
